### PR TITLE
doc: Added External flash memory and Modem trace flash backend

### PR DESCRIPTION
**Pull request by divipillai**
Added following documentation:
External flash memory
Modem trace flash backend

CIA-929 and CIA-928

### CODE REVIEW

**Review by eivindj-nordic — COMMENTED** [review]
Looks good!

File: doc/nrf/working_with_nrf/nrf91/nrf9160.rst [hunk 1]
```diff
@@ -286,6 +283,92 @@ For example, when building a non-secure application for nRF9160 DK v1.0.0, use `
...
 
+.. _nrf9160_external_flash:
+
+External flash memory
+*********************
+
+The nRF9160 DK version 0.14.0 and later versions contain an 8 MB external flash (MX25R6435F).
+The external flash can be used to store data that does not fit on the internal flash.
+To free up GPIO when the external flash is not needed, the nRF9160 DK board controller (nRF52840 SoC) can control a switch to enable or disable routing between the external flash and the nRF9160 SIP.
+See `nRF9160 DK board controller <nRF9160 DK board control section in the nRF9160 DK User Guide_>`_ and `External memory <External memory section in the nRF9160 DK User Guide_>`_ sections in the nRF9160 DK user guides for more details.
```
Comment (wiba-nordic) [inline doc/nrf/working_with_nrf/nrf91/nrf9160.rst:294]:
```suggestion

See `nRF9160 DK board controller <nRF9160 DK board control section in the nRF9160 DK User Guide_>`_ and `External memory <External memory section in the nRF9160 DK User Guide_>`_ sections in the nRF9160 DK user guides for more details.
```

File: doc/nrf/working_with_nrf/nrf91/nrf9160.rst [hunk 2]
```diff
@@ -286,6 +283,92 @@ For example, when building a non-secure application for nRF9160 DK v1.0.0, use `
...
+
+Programming the board controller firmware
+=========================================
+
+To use the external flash memory on the nRF9160 DK v0.14.0 or later versions, you must program the board controller firmware v2.0.1 or higher by completing the following steps:
```
Comment (wiba-nordic) [inline doc/nrf/working_with_nrf/nrf91/nrf9160.rst:299]: Should this bit use `includes/external_flash_nrf91.txt`?

File: doc/nrf/working_with_nrf/nrf91/nrf9160.rst [hunk 3]
```diff
@@ -286,6 +283,92 @@ For example, when building a non-secure application for nRF9160 DK v1.0.0, use `
...
+   By default, this enables pin routing to external flash.
+
+..
+   The new board controller firmware has been updated to enable routing by default (new DK versions?).
+..
+
+Adding the configuration and devicetree option
+===============================================
+
+The application also needs configuration and devicetree options to use the external flash.
```
Comment (wiba-nordic) [inline doc/nrf/working_with_nrf/nrf91/nrf9160.rst:313]:
```suggestion
To use external flash, the application also needs to enable configuration and devicetree options.
```
Not quite happy with this version either, but it at least can't be read as devicetree options using external flash.

File: doc/nrf/working_with_nrf/nrf91/nrf9160.rst [hunk 4]
```diff
@@ -286,6 +283,92 @@ For example, when building a non-secure application for nRF9160 DK v1.0.0, use `
...
+   &mx25r64 {
+       mxicy,mx25r-power-mode = "high-performance";
+   };
+
+For more information about devicetree overlays, see :ref:`zephyr:use-dt-overlays`.
+
+Using Partition Manager
+=======================
+
+If your application was built using |NCS|, you must define partitions using :ref:`partition_manager`.
```
Comment (wiba-nordic) [inline doc/nrf/working_with_nrf/nrf91/nrf9160.rst:355]:
```suggestion
If your application was built using the |NCS|, you must define partitions using :ref:`partition_manager`.
```

File: doc/nrf/working_with_nrf/nrf91/nrf9160.rst [hunk 5]
```diff
@@ -286,6 +283,92 @@ For example, when building a non-secure application for nRF9160 DK v1.0.0, use `
...
+The built-in partition definitions can be found in the :file:`nrf/subsys/partition_manager` folder, and the file names start with ``pm.yml``.
+The files which has ``external_flash`` as their region will have support for storing partitions in the external flash.
```
Comment (wiba-nordic) [inline doc/nrf/working_with_nrf/nrf91/nrf9160.rst:357]:
```suggestion
The files which have ``external_flash`` as their region will have support for storing partitions in the external flash.
```

**Review by wiba-nordic — CHANGES_REQUESTED** [review]

File: doc/nrf/working_with_nrf/nrf91/nrf9160.rst [hunk 6]
```diff
@@ -286,6 +283,92 @@ For example, when building a non-secure application for nRF9160 DK v1.0.0, use `
...
+
+To use the external flash memory on the nRF9160 DK v0.14.0 or later versions, you must program the board controller firmware v2.0.1 or higher by completing the following steps:
+
+#. Download the nRF9160 DK board controller firmware from the `nRF9160 DK downloads`_ page.
+#. Make sure the **PROG/DEBUG SW10** switch on the nRF9160 DK is set to **nRF52**.
+#. Program the board controller firmware (:file:`nrf9160_dk_board_controller_fw_2.0.1.hex`) using the `Programmer app <Programming a Development Kit_>`_ in nRF Connect for Desktop.
+   By default, this enables pin routing to external flash.
+
+..
+   The new board controller firmware has been updated to enable routing by default (new DK versions?).
```
[the remark this answers is not in the document]
Comment (divipillai) [reply doc/nrf/working_with_nrf/nrf91/nrf9160.rst:309]: Isn't this sentence covered in the above points. Is it okay to remove this sentence.
Or, do we need add any other details (as a note) that the user needs to be made aware of?

cc: @gregersrygg.

File: doc/nrf/libraries/modem/nrf_modem_lib.rst [hunk 7]
```diff
@@ -121,6 +122,40 @@ If the difference in the values of the :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRA
...
+* :kconfig:option:`CONFIG_HEAP_MEM_POOL_SIZE` = ``2048``
+* :kconfig:option:`CONFIG_MAIN_STACK_SIZE` = ``4096``
+* :kconfig:option:`CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE` = ``4096``
+* :kconfig:option:`CONFIG_NRF_MODEM_LIB_TRACE_STACK_SIZE` = ``4096``
+
+The modem trace flash backend has some additional configuration options:
+
+* :kconfig:option:`CONFIG_FCB` - required for the flash circular buffer used in the backend.
+* :kconfig:option:`ONFIG_NRF_MODEM_LIB_TRACE_BACKEND_FLASH_PARTITION_SIZE` -  defines the space to be used for the modem trace partition.
+  The maximum available space is 0x800000 (8 MB).
```
[the remark this answers is not in the document]
Comment (divipillai) [reply doc/nrf/libraries/modem/nrf_modem_lib.rst]: Rephrased a bit. Please check.

**Review by wiba-nordic — APPROVED** [review]

File: doc/nrf/working_with_nrf/nrf91/nrf9160.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (eivindj-nordic) [reply doc/nrf/working_with_nrf/nrf91/nrf9160.rst]: Yes, I think the above should cover it.

**Review by eivindj-nordic — APPROVED** [review]

**Pull request merged**